Source/WebCore/rendering/RenderBlockFlow.cpp

```diff
@@ -1193,12 +1214,33 @@ void RenderBlockFlow::determineLogicalLeftPositionForChild(RenderBox& child, App
     // If the child is being centred then the margin calculated to do that has factored in any offset required to
     // avoid floats, so use it if necessary.
 
-    if (style().textAlign() == TextAlignMode::WebKitCenter || child.style().marginStartUsing(&style()).isAuto())
+    if (style().textAlign() == TextAlignMode::WebKitCenter || child.style().marginStartUsing(&style()).isAuto()) {
         newPosition = std::max(newPosition, positionToAvoidFloats + childMarginStart);
-    else if (positionToAvoidFloats > initialStartPosition)
+        setLogicalLeftForChild(child, style().isLeftToRightDirection() ? newPosition : totalAvailableLogicalWidth - newPosition - logicalWidthForChild(child), applyDelta);
+        return;
+    }
+
+    if (positionToAvoidFloats > initialStartPosition)
         newPosition = std::max(newPosition, positionToAvoidFloats);
+    newPosition = computeSelfAlignedLogicalLeftPositionForChild(newPosition);
+    setLogicalLeftForChild(child, newPosition, applyDelta);
+}
 
-    setLogicalLeftForChild(child, style().isLeftToRightDirection() ? newPosition : totalAvailableLogicalWidth - newPosition - logicalWidthForChild(child), applyDelta);
+void RenderBlockFlow::applyStretchAlignmentToChildIfNeeded(RenderBox& child)
```

Comment:
The logic here should probably eventually be refactored into a common function since we do something similar in `RenderGrid` and `RenderFlexibleBox`